File: Contents/InitialConsonantSearch/InitialConsonantSearch/Main/TEST/LineSettingViewController.m
//
//  ChartSettingViewController.m
//  ChartTypeTest
//
//  Created by Kwan Hyun Son on 2023/09/01.
//

#import "LineSettingViewController.h"
#import "ConfigCommonHeaderCell.h"
#import "ToolSettingViewModel.h"
#import "LineSettingViewModel.h"
#import "LineSettingCell.h"
#import <IosKit/IosKit.h>

@interface LineSettingViewController () <UITableViewDataSource, UITableViewDelegate, UIColorPickerViewControllerDelegate>
@property (nonatomic, strong) UIColorPickerViewController *colorPickerViewController;
@property (nonatomic, copy, nullable) void (^colorPickerCompletion)(UIColor *);
@end

@implementation LineSettingViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    NSString *identifier = NSStringFromClass([ConfigCommonHeaderCell class]);
    UINib *tableHeaderNib = [UINib nibWithNibName:identifier bundle:[NSBundle mainBundle]];
    [self.tableView registerNib:tableHeaderNib forHeaderFooterViewReuseIdentifier:identifier];
    self.tableView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    self.tableView.contentInset = UIEdgeInsetsZero;
    self.tableView.sectionHeaderTopPadding = 0.0;
    self.tableView.rowHeight = UITableViewAutomaticDimension; // 셀프 사이즈 셀
    self.tableView.estimatedRowHeight = 100.0;
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    
    self.colorPickerViewController = [UIColorPickerViewController new];
    self.colorPickerViewController.supportsAlpha = NO;
    self.colorPickerViewController.delegate = self;
    self.colorPickerViewController.modalPresentationStyle = UIModalPresentationPopover;
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
}

- (void)setViewModel:(LineSettingViewModel *)viewModel {
    _viewModel = viewModel;
    self.navigationItem.title = viewModel.mainTitle;
}

#pragma mark - <UITableViewDataSource>
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [self.viewModel numberOfSections];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.viewModel numberOfRowsInSection:section];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    DTOLineSetting *setting = [self.viewModel cellModelForIndexPath:indexPath];
    LineSettingCell *cell =
        [tableView dequeueReusableCellWithIdentifier:setting.identifier forIndexPath:indexPath];
    cell.data = setting;
    return cell;
}

#pragma mark - <UITableViewDelegate>
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 44.0;
    // 
    // return 27.0; // 27.204081632653061
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    ConfigCommonHeaderCell *header = [tableView dequeueReusableHeaderFooterViewWithIdentifier:NSStringFromClass([ConfigCommonHeaderCell class])];
    UIBackgroundConfiguration *backgroundConfig = [UIBackgroundConfiguration listPlainHeaderFooterConfiguration];
    [backgroundConfig setBackgroundColor: [UIColor systemGray6Color]];
    [header setBackgroundConfiguration: backgroundConfig];
    header.titleLabel.text = self.viewModel.sectionTitles[section];
    return header;
    //
    // UITableViewHeaderFooterView *header = [tableView dequeueReusableHeaderFooterViewWithIdentifier:@"UITableViewHeaderFooterView"];
    // header.backgroundView.backgroundColor = [UIColor systemRedColor];
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(LineSettingCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
}

#pragma mark - <UIColorPickerViewControllerDelegate>

#if __IPHONE_OS_VERSION_MIN_REQUIRED >= 150000 // Deployment Target 이 15.0이다. 기계가 15 이상부터 다 들어온다.

- (void)colorPickerViewController:(UIColorPickerViewController *)viewController
                   didSelectColor:(UIColor *)color
                     continuously:(BOOL)continuously API_AVAILABLE(ios(15.0)) {
    UIColor *chosenColor = viewController.selectedColor; // User has chosen a color.
    if (self.colorPickerCompletion != nil) {
        self.colorPickerCompletion(chosenColor);
    }
    
    if (continuously == NO) {
        if (self.traitCollection.horizontalSizeClass != UIUserInterfaceSizeClassCompact) {
            [viewController dismissViewControllerAnimated:YES completion:^{
                NSLog(@"chosenColor %@", chosenColor);
            }];
        }
    }
    //
    // Dismiss the color picker if the conditions are right:
    // 1) User is not doing a continous pick (tap and drag across multiple colors).
    // 2) Picker is presented on a non-compact device.
    //
    // Use the following check to determine how the color picker was presented (modal or popover).
    // For popover, we want to dismiss it when a color is locked.
    // For modal, the picker has a close button.
    //
}

#else

// NS_DEPRECATED_IOS(14_0, 15_0) Color. returned from the color picker - iOS 14.x and earlier.
- (void)colorPickerViewControllerDidSelectColor:(UIColorPickerViewController *)viewController  {
    UIColor *chosenColor = viewController.selectedColor;
    if (self.colorPickerCompletion != nil) {
        self.colorPickerCompletion(chosenColor);
    }
    if (self.traitCollection.horizontalSizeClass != UIUserInterfaceSizeClassCompact) {
        [viewController dismissViewControllerAnimated:YES completion:^{
            NSLog(@"chosenColor %@", chosenColor);
        }];
    }
    //
    // Use the following check to determine how the color picker was presented (modal or popover).
    // For popover, we want to dismiss it when a color is locked.
    // For modal, the picker has a close button.
}

#endif

- (void)colorPickerViewControllerDidFinish:(UIColorPickerViewController *)viewController {
    // 명시적으로 X 버튼을 눌렀을 때, 실행된다.
}

#pragma mark - Actions

- (IBAction)titleButtonClicked:(UIButton *)sender { // 체크버튼
    DTOLineSetting *data = [self dataForSender:sender];
    data.selected = !data.selected;
    LineSettingCell *cell = [self cellForSender:sender];
    cell.data = data;
    [self synchronize];
}

- (IBAction)colorButtonClicked:(UIButton *)sender { // 칼라버튼
    __weak __typeof(self) weakSelf = self;
    DTOLineSetting *data = [self dataForSender:sender];
    self.colorPickerCompletion = ^(UIColor *selectedColor) {
        sender.backgroundColor = selectedColor;
        data.color = selectedColor;
        [weakSelf synchronize];
    };
    self.colorPickerViewController.selectedColor = sender.backgroundColor;
    self.colorPickerViewController.title = data.title;
    UIPopoverPresentationController *popoverPresentationController = self.colorPickerViewController.popoverPresentationController;
    popoverPresentationController.sourceView = sender;
    [self presentViewController:self.colorPickerViewController animated:YES completion:^{}];
}

- (IBAction)toggleValueChanged:(UISwitch *)sender {
    DTOLineSetting *data = [self dataForSender:sender];
    data.toggleOn = sender.isOn;
}

- (IBAction)dropdownBtnValueChanged:(MGUDropdownButton *)sender {
    DTOLineSetting *data = [self dataForSender:sender];
    data.dropBtnSelectedIndex = sender.selectedIndex;
    [self synchronize];
}

- (IBAction)boldButtonClicked:(UIButton *)sender { // 체크버튼
    DTOLineSetting *data = [self dataForSender:sender];
    data.bold = !data.isBold;
    LineSettingCell *cell = [self cellForSender:sender];
    cell.data = data;
}

- (void)synchronize {
    ToolSettingLineType lineType = self.viewModel.lineType;
    if ([lineType isEqualToString:ToolSettingLineTypeShapeEllipse] ||
        [lineType isEqualToString:ToolSettingLineTypeShapeRectangle] ||
        [lineType isEqualToString:ToolSettingLineTypeShapeTriangle]) {
        NSIndexPath *path = [NSIndexPath indexPathForRow:0 inSection:0];
        LineSettingCell *cell = [self.tableView cellForRowAtIndexPath:path];
        if (cell != nil) {
            __weak __typeof(self) weakSelf = self;
            [self.viewModel synchronize:^{
                DTOLineSetting *data = [weakSelf.viewModel cellModelForIndexPath:path];
                cell.data = data;
            }];
        }
    }
}

#pragma mark - Helper
- (LineSettingCell *)cellForSender:(__kindof UIView *)sender {
    NSIndexPath *indexPath = [self.tableView mgrIndexPathOfCellWhereViewExists:sender];
    return [self.tableView cellForRowAtIndexPath:indexPath];
}

- (DTOLineSetting *)dataForSender:(__kindof UIView *)sender {
    NSIndexPath *indexPath = [self.tableView mgrIndexPathOfCellWhereViewExists:sender];
    return [self.viewModel cellModelForIndexPath:indexPath];
}
@end

File: Contents/InitialConsonantSearch/InitialConsonantSearch_Swift/Main/TEST/LineSettingViewController.swift
//
//  ChartSettingViewController.swift
//  ChartType_Swift
//
//  Created by Kwan Hyun Son on 2023/09/08.
//

import UIKit
import IosKit

final class LineSettingViewController: UIViewController {
    
    // MARK: - Property
    
    @IBOutlet weak var tableView: UITableView!
    
    var viewModel: LineSettingViewModel? {
        didSet {
            navigationItem.title = viewModel?.mainTitle
        }
    }
    
    private var colorPickerViewController: UIColorPickerViewController!
    
    private var colorPickerCompletion: ((UIColor) -> Void)?

    // MARK: - Override
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let identifier = String(describing: ConfigCommonHeaderCell.self)
        let tableHeaderNib = UINib(nibName: identifier, bundle: nil)
        tableView.register(tableHeaderNib, forHeaderFooterViewReuseIdentifier: identifier)
        
        tableView.contentInsetAdjustmentBehavior = .never
        tableView.contentInset = .zero
        tableView.sectionHeaderTopPadding = 0.0
        tableView.rowHeight = UITableView.automaticDimension
        tableView.estimatedRowHeight = 100.0
        tableView.dataSource = self
        tableView.delegate = self
        
        colorPickerViewController = UIColorPickerViewController()
        colorPickerViewController.supportsAlpha = false
        colorPickerViewController.delegate = self
        colorPickerViewController.modalPresentationStyle = .popover
    }

    // MARK: - Actions
    
    @IBAction private func titleButtonClicked(_ sender: UIButton) {
        guard let data = dataForSender(sender)
        else {
            return
        }
        data.isSelected = !data.isSelected
        if let cell = cellForSender(sender) {
            cell.data = data
        }
        synchronize()
    }

    @IBAction private func colorButtonClicked(_ sender: UIButton) {
        guard let data = dataForSender(sender)
        else {
            return
        }
        colorPickerCompletion = { [weak self] selectedColor in
            sender.backgroundColor = selectedColor
            data.color = selectedColor
            self?.synchronize()
        }
        colorPickerViewController.selectedColor = sender.backgroundColor ?? UIColor.clear
        colorPickerViewController.title = data.title
        if let popoverPresentationController = colorPickerViewController.popoverPresentationController {
            popoverPresentationController.sourceView = sender
        }
        present(colorPickerViewController, animated: true, completion: nil)
    }

    @IBAction private func toggleValueChanged(_ sender: UISwitch) {
        guard let data = dataForSender(sender)
        else {
            return
        }
        data.isToggleOn = sender.isOn
    }

    @IBAction private func dropdownBtnValueChanged(_ sender: SKUDropdownButton) {
        guard let data = dataForSender(sender)
        else {
            return
        }
        data.dropBtnSelectedIndex = sender.selectedIndex
        synchronize()
    }

    @IBAction private func boldButtonClicked(_ sender: UIButton) {
        guard let data = dataForSender(sender)
        else {
            return
        }
        data.isBold = !data.isBold
        if let cell = cellForSender(sender) {
            cell.data = data
        }
    }

    private func synchronize() {
        let lineType = viewModel?.lineType
        if lineType == ToolSettingLineType.shapeEllipse ||
           lineType == ToolSettingLineType.shapeRectangle ||
           lineType == ToolSettingLineType.shapeTriangle {
            let indexPath = IndexPath(row: 0, section: 0)
            if let cell = tableView.cellForRow(at: indexPath) as? LineSettingCell {
                viewModel?.synchronize { [weak self] in
                    let data = self?.viewModel?.cellModel(for: indexPath)
                    cell.data = data
                }
            }
        }
    }
    
    // MARK: - Helper

    private func cellForSender(_ sender: UIView) -> LineSettingCell? {
        if let indexPath = tableView.skhIndexPathOfCellWhereViewExists(view: sender) {
            return tableView.cellForRow(at: indexPath) as? LineSettingCell
        }
        return nil
    }

    private func dataForSender(_ sender: UIView) -> DTOLineSetting? {
        if let indexPath = tableView.skhIndexPathOfCellWhereViewExists(view: sender),
           let viewModel = viewModel {
            return viewModel.cellModel(for: indexPath)
        }
        return nil
    }
}

extension LineSettingViewController: UITableViewDataSource {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return viewModel?.numberOfSections() ?? 0
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return viewModel?.numberOfRows(in: section) ?? 0
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let viewModel = viewModel else {
            return UITableViewCell()
        }
        let setting = viewModel.cellModel(for: indexPath)
        if let cell = tableView.dequeueReusableCell(withIdentifier: setting.identifier, for: indexPath) as? LineSettingCell {
            cell.data = setting
            return cell
        }
        return UITableViewCell()
    }
}

extension LineSettingViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 27.0
    }

    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let header = tableView.dequeueReusableHeaderFooterView(withIdentifier: String(describing: ConfigCommonHeaderCell.self)) as? ConfigCommonHeaderCell
        var backgroundConfig = UIBackgroundConfiguration.listPlainHeaderFooter()
        backgroundConfig.backgroundColor = .systemGray6
        header?.backgroundConfiguration = backgroundConfig
        header?.titleLabel.text = viewModel?.sectionTitles[section]
        return header
    }

    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
    }
}

@available(iOS 14.0, *)
extension LineSettingViewController: UIColorPickerViewControllerDelegate {
    
    @available(iOS 15.0, *)
    func colorPickerViewController(
        _ viewController: UIColorPickerViewController,
        didSelect color: UIColor,
        continuously: Bool
    ) {
        let chosenColor = viewController.selectedColor
        colorPickerCompletion?(chosenColor)
        
        if !continuously {
            if traitCollection.horizontalSizeClass != .compact {
                viewController.dismiss(animated: true) {
                    print("chosenColor \(chosenColor)")
                }
            }
        }
    }
    
    
    // 명시적으로 X 버튼을 눌렀을 때, 실행된다.
    func colorPickerViewControllerDidFinish(_ viewController: UIColorPickerViewController) {}

    /// iOS 14.0에서 Deprecated - 본 프로젝트는 최소 지원버전이 iOS 12.0
    @available(iOS, introduced: 14.0, deprecated: 15.0)
    func colorPickerViewControllerDidSelectColor(_ viewController: UIColorPickerViewController) {
        let chosenColor = viewController.selectedColor
        colorPickerCompletion?(chosenColor)
        if traitCollection.horizontalSizeClass != .compact {
            viewController.dismiss(animated: true) {
                print("chosenColor \(chosenColor)")
            }
        }
    }
}
